File: frontend/src/types/exam.ts
/**
 * TypeScript type definitions for PMP 2026 Exam Simulation
 * Matches backend Pydantic schemas in backend/app/schemas/exam.py
 */

// ============ Exam Enums ============

export type ExamStatusType = 'in_progress' | 'completed' | 'abandoned';

// ============ Exam Session Types ============

export interface ExamSession {
  id: string; // UUID
  user_id: string; // UUID
  status: ExamStatusType;
  start_time: string; // ISO datetime
  end_time: string | null;
  total_time_seconds: number;
  questions_count: number;
  correct_count: number;
  current_question_index: number;
  time_expired: boolean;
  created_at: string;
  updated_at: string;
}

export interface ExamSessionDetail extends ExamSession {
  remaining_time_seconds: number;
  answered_count: number;
  flagged_count: number;
}

export interface ExamSessionCreate {
  exam_duration_minutes?: number; // Default: 240 for PMP
  total_questions?: number; // Default: 185 for PMP
}

export interface ExamSessionSubmit {
  force_complete?: boolean;
}

// ============ Exam Answer Types ============

export interface ExamAnswer {
  id: string; // UUID
  exam_session_id: string; // UUID
  question_id: number;
  question_index: number;
  selected_answer: string; // 'A' | 'B' | 'C' | 'D'
  is_correct: boolean;
  time_spent_seconds: number;
  is_flagged: boolean;
  created_at: string;
}

export interface ExamAnswerSubmit {
  question_id: number;
  selected_answer: string; // 'A' | 'B' | 'C' | 'D'
  time_spent_seconds?: number;
  is_flagged?: boolean;
}

// ============ Exam Question Types ============

export interface ExamQuestion {
  question_index: number;
  question_id: number;
  question_text: string;
  option_a: string;
  option_b: string;
  option_c: string;
  option_d: string;
  selected_answer: string | null;
  is_correct: boolean | null;
  is_flagged: boolean;
  time_spent_seconds: number;
  domain_name: string | null;
  task_name: string | null;
}

export interface ExamQuestionsList {
  exam_session_id: string;
  questions: ExamQuestion[];
}

// ============ Exam Report Types ============

export interface DomainPerformance {
  correct: number;
  total: number;
  percentage: number;
}

export interface TaskPerformance {
  correct: number;
  total: number;
}

export interface ExamReport {
  id: string; // UUID
  exam_session_id: string; // UUID
  score_percentage: number;
  domain_breakdown: Record<string, DomainPerformance>;
  task_breakdown: Record<string, TaskPerformance> | null;
  recommendations: string[];
  strengths: string[];
  weaknesses: string[];
  created_at: string;
  updated_at: string;
}

// ============ Exam Result Types ============

export interface ExamResult {
  exam_session_id: string;
  score_percentage: number;
  passed: boolean;
  domain_breakdown: Record<string, DomainPerformance>;
  task_breakdown: Record<string, TaskPerformance>;
  time_spent_seconds: number;
  time_expired: boolean;
  questions_count: number;
  correct_count: number;
}

// ============ Exam Session with Report ============

export interface ExamSessionWithReport extends ExamSession {
  report: ExamReport | null;
}

// ============ Exam Resume Types ============

export interface ExamResumeData {
  session: ExamSessionDetail;
  questions: ExamQuestion[];
  current_question: ExamQuestion | null;
}

// ============ API Response Types ============

export type ExamSessionResponse = ExamSession;

export type ExamSessionDetailResponse = ExamSessionDetail;

export type ExamAnswerResponse = ExamAnswer;

export type ExamReportResponse = ExamReport;

export type ExamResultResponse = ExamResult;

export type ExamResumeResponse = ExamResumeData;

// ============ Exam Coach / Behavior Types ============

export type BehaviorPattern =
  | 'normal'
  | 'rushing'
  | 'dwelling'
  | 'panic'
  | 'guessing'
  | 'flagging_spree'
  | 'skipping'
  | 'revisit_loop';

export type CoachingSeverity = 'info' | 'suggestion' | 'warning' | 'urgent';

export type PaceTrajectory = 'ahead' | 'on_track' | 'behind' | 'critical';

export interface CoachingAlert {
  pattern: string;
  severity: CoachingSeverity;
  title: string;
  message: string;
  suggested_action: string | null;
  question_index: number | null;
}

export interface BehaviorMetrics {
  current_pattern: BehaviorPattern;
  engagement_score: number;
  focus_score: number;
  pace_trajectory: PaceTrajectory;
  time_remaining_minutes: number;
  questions_completed: number;
  avg_time_per_question: number;
}

export interface GameTapeEvent {
  event_type: 'answer' | 'flag' | 'revisit' | 'skip' | 'coaching';
  question_index: number;
  timestamp: string;
  time_spent_seconds: number;
  pattern_detected: BehaviorPattern | null;
  coaching_message: string | null;
  domain_name: string | null;
  is_correct: boolean | null;
}

export interface GameTapeResponse {
  exam_session_id: string;
  events: GameTapeEvent[];
  summary: BehaviorSummary;
}

export interface BehaviorSummary {
  overall_pattern: BehaviorPattern;
  engagement_score: number;
  focus_score: number;
  pace_trajectory: PaceTrajectory;
  total_flags: number;
  max_consecutive_flags: number;
  question_revisits: number;
  avg_time_per_question: number;
  fastest_question: number;
  slowest_question: number;
  coaching_interventions: number;
  pattern_history: PatternHistoryEntry[];
  current_metrics: {
    time_remaining_minutes: number;
    questions_completed: number;
  };
}

export interface PatternHistoryEntry {
  pattern: BehaviorPattern;
  start_q: number;
  end_q: number;
  duration_sec: number;
}

export interface BehaviorMetricsResponse {
  current_pattern: BehaviorPattern;
  engagement_score: number;
  focus_score: number;
  pace_trajectory: PaceTrajectory;
  time_remaining_minutes: number;
  questions_completed: number;
  avg_time_per_question: number;
}

File: backend/app/schemas/exam.py
"""Pydantic schemas for Exam models (ExamSession, ExamAnswer, ExamReport)."""

from __future__ import annotations

from datetime import datetime
from enum import Enum
from typing import TYPE_CHECKING
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, field_validator

if TYPE_CHECKING:
    pass


class ExamStatus(str, Enum):
    """Status of an exam session."""

    IN_PROGRESS = "in_progress"
    COMPLETED = "completed"
    ABANDONED = "abandoned"


# Exam Session Schemas


class ExamSessionCreate(BaseModel):
    """Schema for creating a new exam session."""

    exam_duration_minutes: int | None = Field(
        None,
        ge=30,
        le=300,
        description="Custom exam duration in minutes (default: 240 for PMP)",
    )
    total_questions: int | None = Field(
        None,
        ge=10,
        le=200,
        description="Number of questions (default: 185 for PMP)",
    )
    adaptive_difficulty: bool = Field(
        default=True,
        description="Enable adaptive difficulty based on prior performance",
    )


class ExamSessionResponse(BaseModel):
    """Schema for exam session response."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID = Field(..., description="Exam session UUID")
    user_id: UUID = Field(..., description="User UUID")
    status: str = Field(..., description="Exam status")
    start_time: datetime = Field(..., description="Exam start time")
    end_time: datetime | None = Field(None, description="Exam end time")
    total_time_seconds: int = Field(default=0, description="Total time spent in seconds")
    questions_count: int = Field(..., description="Number of questions")
    correct_count: int = Field(default=0, description="Correct answers count")
    current_question_index: int = Field(default=0, description="Current question index")
    time_expired: bool = Field(default=False, description="Whether time expired")
    created_at: datetime = Field(..., description="Creation timestamp")
    updated_at: datetime = Field(..., description="Last update timestamp")


class ExamSessionDetailResponse(ExamSessionResponse):
    """Schema for exam session detail with remaining time."""

    remaining_time_seconds: int = Field(..., description="Remaining time in seconds")
    answered_count: int = Field(..., description="Number of answered questions")
    flagged_count: int = Field(..., description="Number of flagged questions")


# Exam Answer Schemas


class ExamAnswerSubmit(BaseModel):
    """Schema for submitting an exam answer."""

    question_id: int = Field(..., gt=0, description="Question ID")
    selected_answer: str = Field(
        ...,
        min_length=1,
        max_length=1,
        description="Selected answer (A, B, C, or D)",
    )
    time_spent_seconds: int = Field(
        default=0,
        ge=0,
        le=600,
        description="Time spent on this question in seconds",
    )
    is_flagged: bool = Field(
        default=False,
        description="Whether to flag this question for review",
    )

    @field_validator("selected_answer")
    @classmethod
    def validate_answer(cls, v: str) -> str:
        """Validate that selected_answer is A, B, C, or D."""
        if v.upper() not in ["A", "B", "C", "D"]:
            raise ValueError("selected_answer must be A, B, C, or D")
        return v.upper()


class ExamAnswerResponse(BaseModel):
    """Schema for exam answer submission response."""

    id: UUID = Field(..., description="Answer record UUID")
    exam_session_id: UUID = Field(..., description="Exam session UUID")
    question_id: int = Field(..., description="Question ID")
    question_index: int = Field(..., description="Question index in exam")
    selected_answer: str = Field(..., description="Selected answer")
    is_correct: bool = Field(..., description="Whether answer was correct")
    time_spent_seconds: int = Field(..., description="Time spent in seconds")
    is_flagged: bool = Field(..., description="Whether flagged for review")
    created_at: datetime = Field(..., description="Creation timestamp")


class ExamQuestionItem(BaseModel):
    """Schema for a question in an exam session."""

    question_index: int = Field(..., description="Question index")
    question_id: int = Field(..., description="Question ID")
    question_text: str = Field(..., description="Question text")
    option_a: str = Field(..., description="Option A")
    option_b: str = Field(..., description="Option B")
    option_c: str = Field(..., description="Option C")
    option_d: str = Field(..., description="Option D")
    selected_answer: str | None = Field(None, description="User's selected answer")
    is_correct: bool | None = Field(None, description="Whether answer was correct")
    is_flagged: bool = Field(default=False, description="Whether flagged for review")
    time_spent_seconds: int = Field(default=0, description="Time spent in seconds")
    domain_name: str | None = Field(None, description="Domain name")
    task_name: str | None = Field(None, description="Task name")


class ExamQuestionsResponse(BaseModel):
    """Schema for exam session questions list."""

    exam_session_id: UUID = Field(..., description="Exam session UUID")
    questions: list[ExamQuestionItem] = Field(..., description="List of questions")


# Exam Completion Schemas


class ExamSessionSubmit(BaseModel):
    """Schema for submitting/completing an exam session."""

    force_complete: bool = Field(
        default=False,
        description="Force complete even if not all questions answered",
    )


class ExamResultResponse(BaseModel):
    """Schema for exam completion result."""

    exam_session_id: UUID = Field(..., description="Exam session UUID")
    score_percentage: float = Field(..., description="Overall score percentage")
    passed: bool = Field(..., description="Whether exam was passed")
    domain_breakdown: dict[str, dict] = Field(
        ...,
        description="Performance breakdown by domain",
    )
    task_breakdown: dict[int, dict] = Field(
        ...,
        description="Performance breakdown by task",
    )
    time_spent_seconds: int = Field(..., description="Total time spent in seconds")
    time_expired: bool = Field(..., description="Whether time expired")
    questions_count: int = Field(..., description="Total questions")
    correct_count: int = Field(..., description="Correct answers")


# Exam Report Schemas


class ExamReportResponse(BaseModel):
    """Schema for exam report response."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID = Field(..., description="Report UUID")
    exam_session_id: UUID = Field(..., description="Associated exam session UUID")
    score_percentage: float = Field(..., description="Overall score percentage")
    domain_breakdown: dict[str, dict] = Field(
        ...,
        description="Domain performance breakdown",
    )
    task_breakdown: dict[int, dict] | None = Field(
        None,
        description="Task performance breakdown",
    )
    recommendations: list[str] = Field(..., description="Study recommendations")
    strengths: list[str] = Field(..., description="Areas of strength")
    weaknesses: list[str] = Field(..., description="Areas needing improvement")
    created_at: datetime = Field(..., description="Report creation timestamp")
    updated_at: datetime = Field(..., description="Last update timestamp")


class ExamSessionWithReportResponse(ExamSessionResponse):
    """Schema for exam session including report."""

    report: ExamReportResponse | None = Field(None, description="Exam report")


# Resume Schemas


class ExamResumeResponse(BaseModel):
    """Schema for resuming an exam session."""

    session: ExamSessionDetailResponse = Field(..., description="Exam session details")
    questions: list[ExamQuestionItem] = Field(..., description="Exam questions")
    current_question: ExamQuestionItem | None = Field(
        None,
        description="Current question to resume from",
    )
